### PR TITLE
#5: Added skipping builds for unaffected folders

### PR DESCRIPTION
**Pull request by tomuben**

### CODE REVIEW

File: exasol_script_languages_container_ci/lib/__init__.py [hunk 1]
```diff
@@ -45,3 +47,20 @@ def print_file(filename: Path, writer: Callable[[str], None]):
...
         writer(f.read())
+
+
+def get_files_of_last_commit() -> List[str]:
+    """
+    Returns the files of the last commit of the repo in the cwd.
+    """
+    repo = Repo()
+    commit = repo.head.commit
+    return list(commit.stats.files.keys())
```
Comment (Nicoretti) [inline exasol_script_languages_container_ci/lib/__init__.py:58]: Generally it is better to return iterators and generators, and let the consumer/user of the function decide if a simple iteration does the job or if he/she needs a materialization and therefore e.g. call the list constructor. 

**TL;DR:** Generators/Iterators are usually way more memory efficient and save memory allocation etc. and threfore often even result in overall better performance.

File: test/fixtures.py [hunk 2]
```diff
@@ -37,3 +45,15 @@ def patch_printfile():
...
+@pytest.fixture(autouse=True)
+def patch_get_files_of_last_commit():
+    """
+    This overwrites automatically function
+    "exasol_script_languages_container_ci.lib.get_files_of_last_commit" because within the UnitTests
+    we do not have a git repository. We can't return an empty list, because this would make the CI build skip.
+    """
+    with patch("exasol_script_languages_container_ci.lib.get_files_of_last_commit",
+               MagicMock(return_value=["src/udfclient.cpp"])):
+        yield
```
Comment (Nicoretti) [inline test/fixtures.py:59]: missing newline

File: test/test_ignore_folders.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+from _pytest.tmpdir import TempPathFactory
+
+from exasol_script_languages_container_ci.lib.ci import check_if_need_to_build
+from test.fixtures import tmp_test_dir
+import git
+
+
+def commit_files(repo: git.Repo, repo_path: Path, files_to_commit: List[str]):
+    for file in files_to_commit:
+        Path(repo_path / file).parent.mkdir(parents=True, exist_ok=True)
```
Comment (Nicoretti) [inline test/test_ignore_folders.py:15]:
```suggestion
        (repo_path / file).parent.mkdir(parents=True, exist_ok=True)
```

File: test/fixtures.py (hunk 2 above)
Comment (tomuben) [reply test/fixtures.py:59]: Done

File: test/test_ignore_folders.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+from exasol_script_languages_container_ci.lib.ci import check_if_need_to_build
+from test.fixtures import tmp_test_dir
+import git
+
+
+def commit_files(repo: git.Repo, repo_path: Path, files_to_commit: List[str]):
+    for file in files_to_commit:
+        (repo_path / file).parent.mkdir(parents=True, exist_ok=True)
+        open(repo_path / file, 'w').close()
```
Comment (Nicoretti) [inline test/test_ignore_folders.py:16]: what is this line for? Why open and do nothing?

**Review by Nicoretti — CHANGES_REQUESTED** [review]

File: exasol_script_languages_container_ci/lib/__init__.py (hunk 1 above)
Comment (tomuben) [reply exasol_script_languages_container_ci/lib/__init__.py]: Done

File: test/test_ignore_folders.py (hunk 4 above)
Comment (tomuben) [reply test/test_ignore_folders.py:16]: I added a comment. I hope it's now clearer what's going on...

**Review by Nicoretti — COMMENTED** [review]
Added a couple of small notes

File: .github/workflows/check_ci.yaml [hunk 5]
```diff
@@ -1,6 +1,10 @@
 name: Run Unit Tests
```
Comment (tkilias) [inline .github/workflows/check_ci.yaml:2]: maybe rename file to something with test

File: exasol_script_languages_container_ci/lib/ci.py [hunk 6]
```diff
@@ -36,20 +60,28 @@ def ci(ctx: click.Context,
...
-    ci_push(ctx, flavor_path=flavor_path,
-         target_docker_repository=docker_build_repository, target_docker_tag_prefix=commit_sha,
-         docker_user=docker_user, docker_password=docker_password)
-    ci_push(ctx, flavor_path=flavor_path,
-         target_docker_repository=docker_build_repository, target_docker_tag_prefix="",
-         docker_user=docker_user, docker_password=docker_password)
-
-    if push_to_public_cache:
+
+    need_to_run = rebuild or check_if_need_to_build(config_file, flavor)
```
Comment (tkilias) [inline exasol_script_languages_container_ci/lib/ci.py:64]: what is with release r is this handle by master

**Review by tkilias — CHANGES_REQUESTED** [review]

File: exasol_script_languages_container_ci/lib/__init__.py [hunk 7]
```diff
@@ -1,5 +1,7 @@
+import json
+from contextlib import contextmanager
 from pathlib import Path
-from typing import Callable
+from typing import Callable, Iterable
 from inspect import cleandoc
```
[the remark this answers is not in the document]
Comment (tomuben) [reply exasol_script_languages_container_ci/lib/__init__.py:6]: All those global functions are used by the commands in this module.

File: exasol_script_languages_container_ci/lib/ci.py (hunk 6 above)
Comment (tomuben) [reply exasol_script_languages_container_ci/lib/ci.py:64]: The relase build will be implemented in another ticket.
